bfdd/bfd_packet.c

```diff
@@ -1219,18 +1506,41 @@ int bp_udp6_mhop(const struct vrf *vrf)
 	return sd;
 }
 
+//# tcpdump -dd udp dst port 3785
+struct sock_filter my_filterudp[] = {
+	{0x28, 0, 0, 0x0000000c}, {0x15, 0, 8, 0x00000800},
+	{0x30, 0, 0, 0x00000017}, {0x15, 0, 6, 0x00000011},
+	{0x28, 0, 0, 0x00000014}, {0x45, 4, 0, 0x00001fff},
+	{0xb1, 0, 0, 0x0000000e}, {0x48, 0, 0, 0x00000010},
+	{0x15, 0, 1, 0x00000ec9}, {0x6, 0, 0, 0x00040000},
+	{0x6, 0, 0, 0x00000000},
+};
+
+#define my_filterudp_length 11
+
 int bp_echo_socket(const struct vrf *vrf)
 {
 	int s;
 
 	frr_with_privs(&bglobal.bfdd_privs) {
-		s = vrf_socket(AF_INET, SOCK_DGRAM, 0, vrf->vrf_id, vrf->name);
+		s = vrf_socket(AF_PACKET, SOCK_RAW, ETH_P_IP, vrf->vrf_id,
+			       vrf->name);
 	}
 	if (s == -1)
 		zlog_fatal("echo-socket: socket: %s", strerror(errno));
 
-	bp_set_ipopts(s);
-	bp_bind_ip(s, BFD_DEF_ECHO_PORT);
+	struct sock_fprog pf;
+	struct sockaddr_ll sll;
+
+	// adjust filter for socket to only receive ECHO packets
+	pf.filter = my_filterudp;
+	pf.len = my_filterudp_length;
+	setsockopt(s, SOL_SOCKET, SO_ATTACH_FILTER, &pf, sizeof(pf));
```

Comment:
Please put some error checking in this `setsockopt` and the `bind` below so we can get an info/warn message when it fails.

Reply:
done